std/meta/trait.zig

```diff
@@ -0,0 +1,495 @@
+        pub fn trait(comptime _T: type) bool
+        {
+            comptime
+            {
+                const T = meta.UnwrapContainer(_T);
+                const info = @typeInfo(T);
+                const defs = switch(info)
+                {
+                    builtin.TypeId.Struct => |s| s.defs,
+                    builtin.TypeId.Union => |u| u.defs,
+                    builtin.TypeId.Enum => |e| e.defs,
+                    else => return false,
+                };
+
+                for(defs) |def|
+                {
+                    if(mem.eql(u8, def.name, name)) return def.is_pub;
+                }
+                
+                return false;
+            }
+        }
+    };
+    return Closure.trait;
+}
+
+test "hasDef"
+{
+    const TestStruct = struct.
+    {
+        pub const value = u8(16);
+    };
+    
+    const TestStructFail = struct.
+    {
+        const value = u8(16);
+    };
+
+    debug.assert(hasDef("value")(TestStruct));
+    debug.assert(!hasDef("value")(TestStructFail));
+    debug.assert(hasDef("value")(*TestStruct));
+    debug.assert(!hasDef("value")(**TestStructFail));
+    debug.assert(!hasDef("x")(TestStruct));
+    debug.assert(!hasDef("value")(u8));
+}
+
+///
+pub fn hasFn(comptime name: []const u8) TraitFn
+{
+    const Closure = struct.
+    {
+        pub fn trait(comptime T: type) bool
+        {
+            comptime
+            {
+                if(!hasDef(name)(T)) return false;
+                const DefType = @typeOf(@field(T, name));
+                const def_type_id = @typeId(DefType);
+                return def_type_id == builtin.TypeId.Fn;
+            }
+        }
+    };
+    return Closure.trait;
+}
+
+test "hasFn"
+{
+    const TestStruct = struct.
+    {
+        pub fn useless() void {}
+    };
+
+    debug.assert(hasFn("useless")(TestStruct));
+    debug.assert(!hasFn("append")(TestStruct));
+    debug.assert(!hasFn("useless")(u8));
+}
+
+///
+pub fn hasField(comptime name: []const u8) TraitFn
+{
+    const Closure = struct.
+    {
+        pub fn trait(comptime _T: type) bool
+        {
+            comptime
+            {
+                const T = meta.UnwrapContainer(_T);
+                const info = @typeInfo(T);
+                const fields = switch(info)
+                {
+                    builtin.TypeId.Struct => |s| s.fields,
+                    builtin.TypeId.Union => |u| u.fields,
+                    builtin.TypeId.Enum => |e| e.fields,
+                    else => return false,
+                };
+
+                for(fields) |field|
+                {
+                    if(mem.eql(u8, field.name, name)) return true;
+                }
+                
+                return false;
+            }
+        }
+    };
+    return Closure.trait;
+}
+
+test "hasField"
+{
+    const TestStruct = struct.
+    {
+        value: u32,
+    };
+
+    debug.assert(hasField("value")(TestStruct));
+    debug.assert(hasField("value")(*TestStruct));
```

Comment:
I know, that `.` can access through pointers, but Idk if we really wanna "lie" to the user because of this. The pointer does not have any fields. `.` just does a little magic.

Reply:
In generic code, the user probably doesn't care if they were passed a pointer or not, as long as they can call/access some needed function/field.

Reply:
Actually, I think you're right. Instead we'll have the user call UnwrapContainer(T) --I am all ears for naming suggestions-- themselves if that's the behavior they want.